### ISSUE TITLE
upgrading from v1.13 to v1.14 causes daemonset's container restart unexpectedly

### ISSUE THREAD

**Issue by gaorong** [kubernetes/kubernetes#78633]
**What happened**:
I have a kubernetes cluster with version  v1.10.2 and want to upgrade to v1.14.2 step by step, without the use of cluster lifecycle management tools like kubeadm.
when upgrading from v1.13.2 to v1.14.2, some daemonset's container restart unexpectedly. 

A new `controllerrevision` was created at this time,  with the only difference from the previous one was adding `spec.containers[0].securityContext.procMount: Default`.

`securityContext.procMount` feature was added in v1.12 by #64283, but it's weird to have an impact at version v1.14.

**What you expected to happen**:
do not restart daemonset's container when upgrading 

**How to reproduce it (as minimally and precisely as possible)**:
1. create a daemonset:
```yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  annotations:
  labels:
    name: nvidia-device-plugin-ds
  name: nvidia-device-plugin-daemonset
spec:
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      name: nvidia-device-plugin-ds
  template:
    metadata:
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ""
      creationTimestamp: null
      labels:
        name: nvidia-device-plugin-ds
    spec:
      containers:
      - image: nvidia/k8s-device-plugin:1.11
        imagePullPolicy: IfNotPresent
        name: nvidia-device-plugin-ctr
        resources: {}
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /var/lib/kubelet/device-plugins
          name: device-plugin
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      terminationGracePeriodSeconds: 30
      tolerations:
      - key: CriticalAddonsOnly
        operator: Exists
      - effect: NoSchedule
        key: nvidia.com/gpu
        operator: Exists
      volumes:
      - hostPath:
          path: /var/lib/kubelet/device-plugins
          type: ""
        name: device-plugin
```

2. upgrade kubernetes component from v1.10.2 (or v1.11.2) to v1.14.2

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by gaorong** [comment]
/sig apps
/sig cluster-lifecycle

@janetkuo @k82cn @liggitt @luxas

**Comment by neolit123** [comment]
> upgrade kubernetes component from v1.10.2 (or v1.11.2) to v1.14.2

i'd say it's also important to outline the exact sequence of component upgrades.

**Comment by liggitt** [comment]
likely triggered by https://github.com/kubernetes/kubernetes/pull/72213/files#r243086103

there was a bug in the way defaulting was applied to the field when it was added in v1.12 in https://github.com/kubernetes/kubernetes/pull/64283

However, the fact that a new defaulted field triggers a restart of a daemonset is concerning, and indicates the change detection strategy used is fragile. It seems similar to https://github.com/kubernetes/kubernetes/issues/57167 in that controllers are assuming they can hash or deepequal compare parent objects and spawned pods.

**Comment by gaorong** [comment]
> i'd say it's also important to outline the exact sequence of component upgrades.

@neolit123  first upgrading master component: apiserver, scheduler, controller-manager, (then this issue occurs),  last upgrading add-ons, kubelet, kube-proxy

**Comment by gaorong** [comment]
Agree to what @liggitt said, As a  complete fix, we need to change the logic of "hash or deepequal compare parent objects" to avoid unnecessary restarts of the DaesmonSet pod.

On the other hand, As the controller cannot handle this case for now, could we revert some parts of  https://github.com/kubernetes/kubernetes/pull/72213/files#r243086103 to avoid other people get into this problem?

**Comment by liggitt** [comment]
> could we revert some parts of https://github.com/kubernetes/kubernetes/pull/72213/files#r243086103 to avoid other people get into this problem?

no, that cannot be reverted

**Comment by liggitt** [comment]
By the time a kube-apiserver is upgraded to 1.14, all nodes must be at 1.12+.

Prior to upgrading nodes to 1.12+, all API servers must have been at 1.12+ as well, which would have populated the procMount field in any newly created or updated pods.

Part of upgrading nodes to 1.12 is draining existing pods off of those nodes. That means by the time you get to 1.14, there should be no remaining pods that were originally created against a pre-1.12 API server.

Following the upgrade sequence described at https://kubernetes.io/docs/setup/version-skew-policy/#supported-component-upgrade-order should ensure that the upgrade to 1.14 does not cause any issues for this particular field.

/close

**Comment by liggitt** [comment]
> there should be no remaining pods that were originally created against a pre-1.12 API server.

that is true, but the default applies to a child field of non-pod objects as well, so it is conceivable a ReplicaSet or Deployment could have remained untouched since before 1.12

/reopen
/assign

**Comment by liggitt** [comment]
/milestone v1.15

this is important to resolve in 1.14, and in 1.15 before release if at all possible

**Comment by mml** [comment]
I would like to see test coverage for this, as an upgrade test, since it seems it's a general principle but there might be many ways to violate it.

We should be able to give a resource, and for each resource a set of mutations we think shouldn't cause anything to happen to pods (including the empty mutation: do nothing), and then the test does the upgrade and waits to make sure all the pods were left alone.

**Comment by liggitt** [comment]
agreed. for the moment, I updated the fixtures for our workloads used in the upgrade tests to ones that would have triggered this

**Comment by liggitt** [comment]
fixes opened for 1.12, 1.13, 1.14, and master

**Comment by gaorong** [comment]
thanks @liggitt , I have verified your patch works fine in our cluster environment.

**Comment by liggitt** [comment]
required changes are merged for 1.15, retargeting at 1.14, will close once https://github.com/kubernetes/kubernetes/pull/78883 is merged

**Comment by gaorong** [comment]
close this as #78883 is merged
